drltrace/drltrace_libcalls.cpp

```diff
@@ -0,0 +1,303 @@
+ */
+
+#include "drltrace.h"
+#include <algorithm>
+/****************************************************************************
+ * Routines to support libcalls hashtable
+ */
+
+#define LIBCALLS_TABLE_HASH_BITS 6
+/* We init the following hashtable to be able to get library call arguments when
+ * it is required for printing.
+ */
+static hashtable_t libcalls_table;
+
+static void
+free_args_list(void *p)
+{
+    std::vector<drsys_arg_t *> *args_list = (std::vector<drsys_arg_t *> *) p;
+    std::vector<drsys_arg_t *>::iterator it;
+    for (it = args_list->begin(); it != args_list->end(); ++it)
+        global_free(*it, sizeof(drsys_arg_t), HEAPSTAT_MISC);
+
+    delete args_list;
+}
+
+static void
+init_libcalls_hashtable()
+{
+    hashtable_init_ex(&libcalls_table, LIBCALLS_TABLE_HASH_BITS, HASH_STRING_NOCASE,
+                      false/*!strdup*/, false, free_args_list, NULL, NULL);
+}
+
+void
+libcalls_hashtable_delete()
+{
+    hashtable_delete(&libcalls_table);
+}
+
+std::vector<drsys_arg_t *> *
+libcalls_search(const char *name)
+{
+    return (std::vector<drsys_arg_t *> *)hashtable_lookup(&libcalls_table, (void *)name);
+}
+
+static bool
+libcalls_hashtable_insert(const char *name, std::vector<drsys_arg_t *> *args_list)
+{
+    return hashtable_add(&libcalls_table, (void *)name, (void *)args_list);
+}
+
+/****************************************************************************
+ * Config file parsing routines
+ */
+
+/* The function returns a new drsys_arg_t object allocated on the global heap
+ * that the caller should free.
+ */
+static drsys_arg_t *
+config_parse_type(std::string type_name, uint index)
+{
+
+    ASSERT(!type_name.empty(), "an empty type name was provided");
+
+    drsys_arg_t *arg = (drsys_arg_t *)global_alloc(sizeof(drsys_arg_t), HEAPSTAT_MISC);
+
+    /* sanitize input type, we assume ASCII chars here */
+    std::transform(type_name.begin(), type_name.end(), type_name.begin(), ::toupper);
+    type_name.erase(std::remove(type_name.begin(), type_name.end(), ' '),
+                    type_name.end());
+    type_name.erase(std::remove(type_name.begin(), type_name.end(), '\r'),
+                    type_name.end());
+    type_name.erase(std::remove(type_name.begin(), type_name.end(), '\n'),
+                    type_name.end());
+
+    /* init arg */
+    arg->type = DRSYS_TYPE_UNKNOWN;
+    arg->ordinal = index;
+    arg->size = 0;
+    arg->arg_name = NULL;
+    arg->type_name = NULL;
+    arg->pre = true;
+    arg->mode = DRSYS_PARAM_INLINED; /* considering inlined param by default */
+
+    /* i#1948: Currently, we have only few cross-platform libcalls in the config
+     * file which is possible to use both for Windows and Linux. However, we
+     * need to separate them into two configs and fix CMAKE accordingly.
+     * Moreover, we have to provide two different interfaces for type parsing
+     * in Windows and Linux.
+     */
+
+    /* i#1948: We have to extend a list of supported types here. */
```

Comment:
It is explained in the code style docs https://github.com/DynamoRIO/dynamorio/wiki/Code-Style